### ISSUE TITLE
Detect unreachable code when if-test calls a function returning `NoReturn`

### ISSUE THREAD

**Issue by Andrej730** [astral-sh/ty#3913]
### Summary

```python
from typing import NoReturn


def no_return() -> NoReturn:
    raise TypeError


def test():
    if no_return():
        # Code is not flagged as unreachable
        print("42")
    # Code is not flagged as unreachable
    print('32')

    no_return()
    # Code is flagged as unreachable.
    print('42')
```

[Playground](https://play.ty.dev/799bb1b6-148a-49e4-ae0a-9390c256e339)
[pyright-play](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByYASgKYwCuIKAUHQCanBQpgD6I5VKAFAJRQAtAD5CJLtQBcNKLKggAhkgDOpKABVEpAKIhwIOjUbMYpZTH7S5mZqw4TefK9bkBiKAGEwjTMpZh8YAAbBTQ0UnooBT8KFE4FAGMACwUAIyDSGRdZBBBUCwAiABYAJgK%2BLNl3Lx8Vf0CQsIiomLjSRJT0zOtc-J4AcgBmEv6Kyv97Smp%2BcervNTrg0PDI6KhY%2BOS0jIA6cd6UC37S0ZogA)

Found this by trying to solve problem similar to https://github.com/astral-sh/ty/issues/2312, trying to use `NoReturn` as a way to at least somewhat statically indicate that `__bool__` is not supported.

### Version

ty 0.0.56

**Comment by carljm** [comment]
We explicitly chose to only handle bare statement-level calls, not nested calls, when fixing #180, because the performance cost of handling all nested calls was too high. We could revisit this and try to add a few more cases -- if-tests seem effectively statement-level and so probably should be added, and wouldn't impact performance much.

**Comment by carljm** [comment]
If we do this, we'd probably also want to handle `while fail()` at least. Also `if await fail()` and `while await fail()`. Probably also `a if fail() else b`.

Other things to consider: `a and fail() or b`? Irrefutable match pattern with `fail()` guard (probably not worth it)?
